src/main/java/mate/academy/service/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,107 @@
+        ShoppingCart shoppingCart = getShoppingCartByCurrentUser();
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    @Transactional
+    public ShoppingCartDto addBookToShoppingCart(CreateCartItemRequestDto requestDto) {
+        ShoppingCart shoppingCart = getShoppingCartByCurrentUser();
+        addOrUpdateCartItem(requestDto, shoppingCart);
+        shoppingCartRepository.save(shoppingCart);
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    @Transactional
+    public ShoppingCartDto updateShoppingCart(Long cartItemId,
+                                              UpdateCartItemDto updateCartItemDto) {
+        ShoppingCart shoppingCart = getShoppingCartByCurrentUser();
+        CartItem existingCartItem = shoppingCart.getCartItems().stream()
+                .filter(item -> item.getId().equals(cartItemId))
+                .findFirst()
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can't find cart item with id:" + cartItemId));
+        cartItemMapper.updateCartItemFromDto(updateCartItemDto, existingCartItem);
+        shoppingCartRepository.save(shoppingCart);
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    public void removeCartItem(Long cartItemId) {
+        ShoppingCart shoppingCart = getShoppingCartByCurrentUser();
+        shoppingCart.getCartItems().removeIf(item -> item.getId().equals(cartItemId));
+        shoppingCartRepository.save(shoppingCart);
+    }
+
+    private Long getCurrentUserId() {
+        UserDetails principal = (UserDetails) SecurityContextHolder.getContext()
+                .getAuthentication()
+                .getPrincipal();
+        return ((User) principal).getId();
+    }
+
+    private ShoppingCart getShoppingCartByCurrentUser() {
+        return shoppingCartRepository.findByUserId(getCurrentUserId());
+    }
+
+    private void addOrUpdateCartItem(CreateCartItemRequestDto requestDto,
+                                     ShoppingCart shoppingCart) {
+        Book book = bookRepository.findById(requestDto.bookId()).orElseThrow(() ->
+                new EntityNotFoundException("Can't find book by id: " + requestDto.bookId())
+        );
+        Optional<CartItem> existingCartItem = shoppingCart.getCartItems().stream()
+                .filter(item -> item.getBook().getId().equals(book.getId()))
+                .findFirst();
+        if (existingCartItem.isPresent()) {
+            int updatedQuantity = existingCartItem.get().getQuantity() + requestDto.quantity();
+            existingCartItem.get().setQuantity(updatedQuantity);
+        } else {
+            CartItem newCartItem = cartItemMapper.toEntity(requestDto);
+            newCartItem.setBook(book);
+            newCartItem.setShoppingCart(shoppingCart);
+            shoppingCart.getCartItems().add(newCartItem);
+            cartItemRepository.save(newCartItem);
+        }
```

Comment:
better avoid if-else construction  
```suggestion
        if (existingCartItem.isPresent()) {
            int updatedQuantity = existingCartItem.get().getQuantity() + requestDto.quantity();
            existingCartItem.get().setQuantity(updatedQuantity);
            return;
        } 
            CartItem newCartItem = cartItemMapper.toEntity(requestDto);
            newCartItem.setBook(book);
            newCartItem.setShoppingCart(shoppingCart);
            shoppingCart.getCartItems().add(newCartItem);
            cartItemRepository.save(newCartItem);
```

Reply:
Thank you.